File: scripts/publish.py
#!/usr/bin/env python
import argparse
import json
import os
import shutil
import subprocess
import sys

ROOT_DIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))


class PackageConfigPatcher:
    def __init__(self, suffix=''):
        self._suffix = suffix
        self._config_path = os.path.join(ROOT_DIR, 'package.json')
        self._orig_config = ''

    def __enter__(self):
        with open(self._config_path, 'r') as f:
            self._orig_config = f.read()
        patched_config = self._create_patched_config(self._orig_config)
        with open(self._config_path, 'w') as f:
            f.write(patched_config)

    def __exit__(self, exc_type, exc_value, traceback):
        with open(self._config_path, 'w') as f:
            f.write(self._orig_config)

    def _create_patched_config(self, config):
        patched_config = json.loads(config)
        if self._suffix:
            patched_config[
                'version'] = patched_config['version'] + '-' + self._suffix
        return json.dumps(patched_config, indent=2)


def parse_args():
    arg_parser = argparse.ArgumentParser()

    arg_parser.add_argument(
        '--dry-run', action='store_true', help='Dry run mode for npm publish')
    arg_parser.add_argument(
        '--tag', '-T', type=str, required=True, help='NPM published tag')
    arg_parser.add_argument(
        'dist_tar_file', action='store', help='dist.tgz created from CI')

    args = arg_parser.parse_args()
    if not args.dist_tar_file:
        arg_parser.print_help()
        sys.exit(1)
    return args


def main():
    args = parse_args()

    workdir = os.path.join(ROOT_DIR, 'build', 'publish')
    if not os.path.exists(workdir):
        os.makedirs(workdir)
    distdir = os.path.join(ROOT_DIR, 'dist')
    dryrun = '--dry-run' if args.dry_run else ''

    print('\n\n========== Publish standard package ==========')
    with PackageConfigPatcher(''):
        if os.path.exists(distdir):
            shutil.rmtree(distdir)
        subprocess.check_call(
            ['tar', '-xf', args.dist_tar_file, '-C', workdir])
        shutil.move(os.path.join(workdir, 'dist'), distdir)
        subprocess.check_call(['npm', 'publish', dryrun, '--tag', args.tag])

    print('\n\n========== Publish unstripped package ==========')
    with PackageConfigPatcher('unstripped'):
        if os.path.exists(distdir):
            shutil.rmtree(distdir)
        subprocess.check_call(
            ['tar', '-xf', args.dist_tar_file, '-C', workdir])
        shutil.move(os.path.join(workdir, 'dist.unstripped'), distdir)
        subprocess.check_call(
            ['npm', 'publish', dryrun, '--tag', args.tag + '-unstripped'])

    shutil.rmtree(workdir)


if __name__ == '__main__':
    main()

File: scripts/publish.js
#!/usr/bin/env node
/*
 * @format
 */

const child_process = require('child_process');
const commander = require('commander');
const fs = require('fs');
const path = require('path');
const rimraf = require('rimraf');
const semver = require('semver');

if (!semver.satisfies(process.versions.node, '>= 10.12.0')) {
  console.log('Please execute this script with node version >= 10.12.0');
  process.exit(1);
}

commander
  .requiredOption('-T, --tag <tag>', 'NPM published tag')
  .arguments('<artifact_zip_file>')
  .option('--dry-run', 'Dry run mode for npm publish')
  .parse(process.argv);

const artifactZipFile = verifyFile(commander.args[0], '<artifact_zip_file>');
const rootDir = path.dirname(__dirname);
const workDir = path.join(rootDir, 'build', 'publish');
const distDir = path.join(rootDir, 'dist');
if (fs.existsSync(workDir)) {
  rimraf.sync(workDir);
}
fs.mkdirSync(workDir, {recursive: true});

child_process.execFileSync('unzip', [artifactZipFile, '-d', workDir]);

// Publish standard package
console.log('\n\n========== Publish standard package ==========');
createPatchedContext(rootDir, '', () => {
  if (fs.existsSync(distDir)) {
    rimraf.sync(distDir);
  }
  fs.renameSync(path.join(workDir, 'dist'), distDir);
  const publishArgs = ['publish', '--tag', commander.tag];
  if (commander.dryRun) {
    publishArgs.push('--dry-run');
  }
  child_process.execFileSync('npm', publishArgs);
});

// Publish unstripped package
// 1. Add suffix in version, e.g. 245459.0.0-unstripped
// 2. Add suffix in tag, e.g. latest-unstripped
// 3. Get unstripped distribution from dist.unstripped/ in CI archive.zip
console.log('\n\n========== Publish unstripped package ==========');
createPatchedContext(rootDir, 'unstripped', () => {
  if (fs.existsSync(distDir)) {
    rimraf.sync(distDir);
  }
  fs.renameSync(path.join(workDir, 'dist.unstripped'), distDir);
  const publishArgs = ['publish', '--tag', `${commander.tag}-unstripped`];
  if (commander.dryRun) {
    publishArgs.push('--dry-run');
  }
  child_process.execFileSync('npm', publishArgs);
});

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------
function verifyFile(filePath, argName) {
  if (filePath == null) {
    console.error(`Error: ${argName} is required`);
    process.exit(1);
  }

  let stat;
  try {
    stat = fs.lstatSync(filePath);
  } catch (error) {
    console.error(error.toString());
    process.exit(1);
  }

  if (!stat.isFile()) {
    console.error(`Error: ${argName} is not a regular file`);
    process.exit(1);
  }

  return filePath;
}

function createPatchedContext(rootDir, versionSuffix, wrappedRunner) {
  const configPath = path.join(rootDir, 'package.json');
  const origConfig = fs.readFileSync(configPath);

  function enter() {
    const patchedConfig = JSON.parse(origConfig);
    if (versionSuffix) {
      patchedConfig.version += '-' + versionSuffix;
    }
    fs.writeFileSync(configPath, JSON.stringify(patchedConfig, null, 2));
  }

  function exit() {
    fs.writeFileSync(configPath, origConfig);
  }

  enter();
  const ret = wrappedRunner.apply(this, arguments);
  exit();
  return ret;
}
